Run flutter tests in preview-dart-2 mode on travis flutter builds.

aam opened the pull request:
cc: @a-siva

aam committed:
Run flutter tests in preview-dart-2 mode on travis flutter builds.

a-siva commented:
The changes look good but I am not sure why you are getting test errors in all the shards. Maybe you should sync up and try the tests again to see what state they end up in.

Hixie commented on .travis.yml:
```diff
@@ -14,6 +15,8 @@ matrix:
   - os: osx
     env: SHARD=docs
+  - os: osx
+    env: SHARD=tests_dart2
```
this excludes it from osx. i think we want it enabled for osx, no?

aam replied:
Was not sure we want to run them on osx, but enabled now.

Hixie commented on dev/bots/test.dart:
```diff
@@ -171,21 +187,21 @@ Future<Null> _runTests() async {
   await _pubRunTest(path.join(flutterRoot, 'packages', 'flutter_tools'));
   await _pubRunTest(path.join(flutterRoot, 'dev', 'bots'));
 
   await _runAllDartTests(path.join(flutterRoot, 'dev', 'devicelab'));
```
can we run dart tests with preview2 too?

aam replied:
Yes, that should be possible.

Hixie commented on dev/bots/test.dart:
```diff
@@ -171,21 +187,21 @@ Future<Null> _runTests() async {
+  await _runFlutterTest(path.join(flutterRoot, 'packages', 'flutter_driver'), options: options);
+  await _runFlutterTest(path.join(flutterRoot, 'packages', 'flutter_test'), options: options);
   await _pubRunTest(path.join(flutterRoot, 'packages', 'flutter_tools'));
   await _pubRunTest(path.join(flutterRoot, 'dev', 'bots'));
```
how about these?

aam replied:
Not sure how `pub run` can run in dart2 mode.

a-siva replied:
We need to file an issue against pub for this.
Currently it has the following options for pub run
Usage: pub run <executable> [args...]
-h, --help            Print this usage information.
-c, --[no-]checked    Enable runtime type checks and assertions.
    --mode            Mode to run transformers in.
                      (defaults to "release" for dependencies, "debug" for entrypoint)

it needs another option here '--preview-dart-2' to be consistent that would run in dart2 mode.

aam replied:
Opened https://github.com/dart-lang/sdk/issues/32214

aam committed:
Merge branch 'master' into run-test-dart2

aam committed:
Run dart2 tests on osx. Run dart tests in dart2.

aam committed:
Fix name camelCase

aam committed:
Default options to empty array, rather than null

aam committed:
Merge branch 'master' into run-test-dart2

aam committed:
Troubleshoot failures

aam committed:
More logging

aam committed:
Merge branch 'master' into run-test-dart2

aam committed:
Troubleshoot: run single test

aam committed:
Troubleshoot: run 15 tests

aam committed:
Troubleshoot: run 15 tests with fix

a-siva commented:
Maybe we should move running under dart2 mode to the Allowed Failure section so that we can land this PR and then fix up the tests.

aam commented:
> Maybe we should move running under dart2 mode to the Allowed Failure section so that we can land this PR and then fix up the tests.

I would have been happy to do that, but I'm seeing some weird timeout issues with running dart2 tests all at once. It looks like machine performance tanks when you run ~30 frontend tests in parallel(15 tests work).

aam committed:
Try limit concurrency to 1

aam committed:
Limit concurrency for preview-dart-2 tests to 4

aam commented:
> > Maybe we should move running under dart2 mode to the Allowed Failure section so that we can land this PR and then fix up the tests.

> I would have been happy to do that, but I'm seeing some weird timeout issues with running dart2 tests all at once. It looks like machine performance tanks when you run ~30 frontend tests in parallel(15 tests work

Limiting concurrency to 4 seems to work fine.

aam committed:
Move dart2 tests to allow_failures section

aam committed:
Reinstate tests_dart_2 shard

aam committed:
Raise concurrency to 8

mraleph commented:
I think we need to investigate reusing a single frontend_server instance for running multiple tests. Spawning a new one per test is likely way too much overhead.

mraleph commented:
FYI: `flutter_driver` test failures should be fixed by  dart-lang/sdk@eb997b281832c27a59e13526bf008014338785c8, I also filed dart-lang/sdk#32185 and dart-lang/sdk#32183 for compilation errors in file and vmservice_client packages.

aam committed:
Merge branch 'master' into run-test-dart2

aam committed:
Merge

aam committed:
Reuse compiler across multiple test runs

aam committed:
Allow to switch entry points when recompiling.

Actually use single compiler to incrementally recompile all tests executed by 'flutter test'.

aam committed:
Remove leftover commented code

a-siva commented on packages/flutter_tools/lib/src/test/flutter_platform.dart:
```diff
@@ -101,16 +101,32 @@ class _FlutterPlatform extends PlatformPlugin {
+      bool isEmpty = compilationQueue.isEmpty;
+      compilationQueue.add(request);
+      // Only trigger processing if queue was empty - i.e. no other requests
+      // we being processed. This effectively enforces "one compilation request at a time".
```
no other requests are currently being processed

aam replied:
Thanks, Done.

a-siva commented on packages/flutter_tools/lib/src/test/flutter_platform.dart:
```diff
@@ -93,7 +100,32 @@ class _FlutterPlatform extends PlatformPlugin {
+          compiler.accept();
+          compiler.reset();
+          request.result.complete(outputPath);
+          // Only remove now when we finished processing the element
```
in what circumstance would we try recompiling the same test again ? I am wondering why is is necessary to remove the request only after we are done with the compilation.

aam replied:
It's not that we are going to recompile the same test again. We only remove the test from the queue after we are done compiling because that way we know that no other compilation request can start before we exit(no await statements from this point on).

aam committed:
Fix comment

aam committed:
Merge branch 'master' into run-test-dart2

a-siva commented:
What is the final outcome of this, do we run the Travis tests using the batch compiler, I thought this would cause timeouts as we would exceed the Travis time budgets.

aam commented:
> What is the final outcome of this, do we run the Travis tests using the batch compiler? I thought this would cause timeouts as we would exceed the Travis time budgets.

Correct, this is now blocked on incremental compiler issue revealed in https://github.com/flutter/flutter/issues/14931.

aam commented:
Blocked on #32368

aam committed:
Merge

aam committed:
Lints

aam committed:
Merge branch 'master' into run-test-dart2

aam commented:
With dart-lang/sdk#32368 fixed, this PR that shares frontend server between tests travis build in preview-dart-2 takes 26m47s, "only" 20s more than non-preview-dart-2 test, so is ready to be submitted I believe.

mraleph commented:
It looks more like 3 minutes slower on Mac, but *8 minutes* slower on Linux if I am not misreading the output. 

<img width="995" alt="build__21078_-_flutter_flutter_-_travis_ci" src="https://user-images.githubusercontent.com/131846/36909942-b130734a-1e3f-11e8-9a9b-6229d424402a.png">

I think this warrants investigation, to see if we can do something about it /cc @jensjoha @kmillikin

aam commented:
as @Hixie pointed out comparing numbers from `Build Jobs` dashboard does not full picture as preview-dart-2 tests don't run all the way to the completion.
Better comparison at this point would be to look at execution of `flutter test` in `packages/flutter` which completes successfully in both configurations:
- `06:13 +2327 ~16: All tests passed!` without preview-dart-2 vs
- `18:28 +2327 ~16: All tests passed!` with preview-dart-2

tvolkert commented:
Looks like this broke internal Google tests - /cc @keertip

tvolkert commented:
And looks like @jason-simmons already has a fix internally

jensjoha commented:
For now I have a few comments:
* Running `flutter test --preview-dart-2` in packages/flutter doesn't work for me at all. The VM crashes because of a deleted file after which all tests time out (after 12 minutes!). I have a fix for that though (fix in dart sdk/pkg/vm; still not uploaded).
* When running `flutter test` my computer uses 100% CPU on all (12) cores, when running `flutter test --preview-dart-2` it uses 100% CPU on 1 core and "a little bit" on the other ones.
* The tests basically does this: recompile, accept, reset, where reset throws away the compiler inside the frontend_server and creates a new one. That doesn't seem ideal.

After fixing whats talked about in bullet-point #1 above I get
* 01:10 +2334 ~16: All tests passed! (`flutter test --local-engine=host_debug`)
* 08:34 +2333 ~16 -1: Some tests failed. (`flutter test --preview-dart-2 --local-engine=host_debug`)
